pkg/spdk/disk/driver.go

```diff
@@ -0,0 +1,88 @@
+package disk
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	spdkclient "github.com/longhorn/go-spdk-helper/pkg/spdk/client"
+	spdktypes "github.com/longhorn/go-spdk-helper/pkg/spdk/types"
+)
+
+type DiskDriver interface {
+	DiskCreate(*spdkclient.Client, string, string, uint64) (string, error)
+	DiskDelete(*spdkclient.Client, string, string) (bool, error)
+	DiskGet(*spdkclient.Client, string, string, uint64) ([]spdktypes.BdevInfo, error)
+}
+
+var (
+	diskDrivers map[string]DiskDriver
+)
+
+func init() {
+	diskDrivers = make(map[string]DiskDriver)
+}
+
+func RegisterDiskDriver(diskDriver string, ops DiskDriver) error {
+	if _, exists := diskDrivers[diskDriver]; exists {
+		return fmt.Errorf("disk driver %s has already been registered", diskDriver)
+	}
+	diskDrivers[diskDriver] = ops
+	return nil
+}
+
+func DiskCreate(spdkClient *spdkclient.Client, diskName, diskPath, diskDriver string, blockSize uint64) (string, error) {
+	driver, ok := diskDrivers[diskDriver]
+	if !ok {
+		return "", fmt.Errorf("disk driver %s is not registered", diskDriver)
+	}
+
+	return driver.DiskCreate(spdkClient, diskName, diskPath, blockSize)
+}
+
+func DiskDelete(spdkClient *spdkclient.Client, diskName, diskPath, diskDriver string) (bool, error) {
+	driver, ok := diskDrivers[diskDriver]
+	if !ok {
+		return false, fmt.Errorf("disk driver %s is not registered", diskDriver)
+	}
+
+	return driver.DiskDelete(spdkClient, diskName, diskPath)
+}
+
+func DiskGet(spdkClient *spdkclient.Client, diskName, diskPath, diskDriver string, timeout uint64) ([]spdktypes.BdevInfo, error) {
```

Comment:
NIT: Can we use `UUID` as the alternative? If `UUID` is not empty, use it then verify other non-empty inputs. Otherwise, fall back to the current implementation.

Reply:
Currently, ddiskGet calls `BdevGetBdevs` which only uses `diskName`. We don't support renaming disk for v2 data engine. Can handle the issue in another improvement ticket?

Reply:
The RPC `bdev_get_bdevs`, which is used by `BdevGetBdevs`, can be called both with `bdev name` and `bdev alias`.
For logical volume, the `bdev name` is the lvol's uuid and the `bdev alias` is the lvol's name.
For other bdevs is the opposite, for example for a malloc bdev the `bdev name` is the malloc's uuid and the `bdev alias` is the malloc's name.

Reply:
As Damiano mentioned, it's doable. But YES I can handle this with a separate ticket.

Reply:
Let's handle it in a separate ticket.
But what's the benefit of UUID in the function?

Reply:
> The RPC `bdev_get_bdevs`, which is used by `BdevGetBdevs`, can be called both with `bdev name` and `bdev alias`. For logical volume, the `bdev name` is the lvol's uuid and the `bdev alias` is the lvol's name. For other bdevs is the opposite, for example for a malloc bdev the `bdev name` is the malloc's uuid and the `bdev alias` is the malloc's name.

The naming rules in spdk are pretty messy...

Reply:
Yes. And I tried to distinguish them in our repos:
`lvolName`: The latter part of the lvol alias. It does not contain the lvs name.
`lvolAlias`: `<lvs name>/<lvol name>`

And there are more potential conventions against the other kinds of naming like Longhorn snapshot name, snasphot lvol name, backup, and so on. If required, we can add some comments or write a doc for the related repos.

Reply:
> what's the benefit of UUID in the function
It's more precise. And it's convention for debugging purpose.